src/waveform/renderers/qopengl/waveformrendererfiltered.cpp

```diff
@@ -0,0 +1,255 @@
+        int visualFrameStop = int(xVisualSampleIndex / 2.0 + maxSamplingRange + 0.5);
+        const int lastVisualFrame = dataSize / 2 - 1;
+
+        // We now know that some subset of [visualFrameStart, visualFrameStop]
+        // lies within the valid range of visual frames. Clamp
+        // visualFrameStart/Stop to within [0, lastVisualFrame].
+        visualFrameStart = math_clamp(visualFrameStart, 0, lastVisualFrame);
+        visualFrameStop = math_clamp(visualFrameStop, 0, lastVisualFrame);
+
+        int visualIndexStart = visualFrameStart * 2;
+        int visualIndexStop = visualFrameStop * 2;
+
+        visualIndexStart = std::max(visualIndexStart, 0);
+        visualIndexStop = std::min(visualIndexStop, dataSize);
+
+        // 3 bands, 2 channels
+        float max[3][2]{};
+
+        for (int i = visualIndexStart; i < visualIndexStop; i += 2) {
+            const WaveformData& waveformData = data[i];
+            const WaveformData& waveformDataNext = data[i + 1];
+
+            const float filteredLow = static_cast<float>(waveformData.filtered.low);
+            const float filteredMid = static_cast<float>(waveformData.filtered.mid);
+            const float filteredHigh = static_cast<float>(waveformData.filtered.high);
+
+            const float nextFilteredLow = static_cast<float>(waveformDataNext.filtered.low);
+            const float nextFilteredMid = static_cast<float>(waveformDataNext.filtered.mid);
+            const float nextFilteredHigh = static_cast<float>(waveformDataNext.filtered.high);
+
+            max[0][0] = math_max(max[0][0], filteredLow);
+            max[0][1] = math_max(max[0][1], nextFilteredLow);
+            max[1][0] = math_max(max[1][0], filteredMid);
+            max[1][1] = math_max(max[1][1], nextFilteredMid);
+            max[2][0] = math_max(max[2][0], filteredHigh);
+            max[2][1] = math_max(max[2][1], nextFilteredHigh);
+        }
+
+        const float fpos = static_cast<float>(pos);
+
+        for (int bandIndex = 0; bandIndex < 3; bandIndex++) {
+            max[bandIndex][0] *= bandGain[bandIndex];
+            max[bandIndex][1] *= bandGain[bandIndex];
+
+            // lines are thin rectangles
+            addRectangle(m_lineValues[bandIndex],
+                    fpos - 0.5f,
+                    halfBreadth - heightFactor * max[bandIndex][0],
+                    fpos + 0.5f,
+                    halfBreadth + heightFactor * max[bandIndex][1]);
+        }
+
+        xVisualSampleIndex += gain;
+    }
+
+    QMatrix4x4 matrix;
+    matrix.ortho(QRectF(0.0,
+            0.0,
+            m_waveformRenderer->getWidth() * devicePixelRatio,
+            m_waveformRenderer->getHeight() * devicePixelRatio));
+    if (m_waveformRenderer->getOrientation() == Qt::Vertical) {
+        matrix.rotate(90.f, 0.0f, 0.0f, 1.0f);
+        matrix.translate(0.f, -m_waveformRenderer->getWidth() * devicePixelRatio, 0.f);
+    }
```

Comment:
second time occuring. I think this deserves a function somewhere in the more abstract base-classes.

Reply:
i put it in a helper function, as there is no suitable common base class for this